Date Range Component

getdave opened the pull request:
## Changes proposed in this Pull Request

Add a new `DateRange` component, extracted from changes originally proposed in [`25762`](https://github.com/Automattic/wp-calypso/pull/25762).

## Testing instructions

This is a new component so please follow the Component Checklist README.

Also:
* View and test the example at http://calypso.localhost:3000/devdocs/design/date-range
* Run the tests `npm run test-client:watch client/components/date-range/test`
* Check the existing DateRange component in the `Activity` section of Calypso - ref: https://github.com/Automattic/wp-calypso/pull/29938/commits/22fcf7f367047efd0d58ebc30f58066e7224bfa4

## Todo

- [x] ~~Switch calendar to month when start/end input receives `focus`~~
- [x] ~~Match styling of range to the range in the Activity log~~
- [x] ~~Test iOS regression re: touch event handlers (see https://github.com/Automattic/wp-calypso/commit/406a5f6226761b5ef3ff17f35f5d42956880ea6e)~~
- [x] ~~Fix tests~~
- [x] ~~Resolve timezone issues highlighted by [this feedback](https://github.com/Automattic/wp-calypso/pull/29938#issuecomment-453569311). Likely need to convert all tests to use `ISO` format for assertions to ensure consistent results~~
- [x] ~~Update color vars as per `p4TIVU-9c1-p2`~~
- [x] ~~Address @mattsherman CR points~~
- [x] ~~Document render props~~
- [x] ~~DRY up date range clamping validation code~~
- [x] ~~Clamping of start/end date props to ensure they conform to `first|lastSelectableDate` prop ranges~~
- [x] ~~https://github.com/Automattic/wp-calypso/pull/25762#pullrequestreview-189859746~~
- [x] ~~https://github.com/Automattic/wp-calypso/pull/25762#pullrequestreview-189774450~~
- [x] ~~https://github.com/Automattic/wp-calypso/pull/25762#discussion_r245709897~~
- [x] ~~https://github.com/Automattic/wp-calypso/pull/29938#discussion_r245840735~~
- [x] ~~https://github.com/Automattic/wp-calypso/pull/29938#discussion_r245838307~~
- [x] ~~Refactor tests~~

spen commented on client/components/date-picker/index.jsx:
```diff
@@ -88,7 +96,7 @@ class DatePicker extends PureComponent {
 	}
 
-	locale() {
+	localeUtils() {
```
Nitpicking a little here but I'd name this `getLocaleUtils` or similar. Generally though, the existing method seems a bit weird :/

getdave replied:
Moved to https://github.com/Automattic/wp-calypso/pull/29981

spen commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,300 @@
+
+	handleInputBlur( e ) {
+		const val = e.target.value;
+		const startOrEnd = e.target.id;
```
This worries me for similar reasons - we're relying on something quite implicit. I think we'd be able to do something similar to the suggestion above.

getdave replied:
👍

getdave replied:
See above

spen commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,300 @@
+				calendarViewDate={ this.momentDateToNative( this.state.startDate ) }
+				disabledDays={ [
+					{
+						after: now, // you can't look at photos from the future!
```
This no longer really applies right? As a general component we should allow for future dates too and consider how best to limit days via props... We could allow `disabledDays` to be passed as props or have set restrictions via props like `allowFutureDates` & `allowPastDates` or similar.

getdave replied:
I think `disabledDays` prop.

getdave replied:
@spen Please see https://github.com/Automattic/wp-calypso/pull/29938/commits/142cc8b2ef58e994bea9b0d3c2fcd0aaf81619d9

getdave commented on client/components/date-picker/index.jsx:
```diff
@@ -120,7 +128,7 @@ class DatePicker extends PureComponent {
 			},
 		};
 
-		return merge( locale, this.props.locale );
```
Moved to https://github.com/Automattic/wp-calypso/pull/29981

getdave commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,300 @@
+				// just in case user doesn't "Apply" and we need to revert
+				// to the original dates
+				if ( ! this.state.oldDatesSaved ) {
+					newState = Object.assign( {}, newState, {
```
Fixed

getdave commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,300 @@
+	handleInputChange( e ) {
+		const val = e.target.value;
+
+		const fromOrTo = e.target.id.includes( 'start' ) ? 'From' : 'To';
```
Love this - thanks @spen. I've used the partial application trick before and you are right it's far more robust.

Will look at implementing tomorrow. Not keen on moving the component back into the main class, but I'm open to being convinced :)

getdave replied:
Now addressed via method suggested

johngodley commented:
This is looking great and feels polished already. It makes a lot of sense to pull this out into a proper component.

One slight thing I noticed is that the placeholder (the button you initially click to open the popover) changes width slightly as you change dates, making the popover move about. It's only a few pixels but you see it if you click between the 1st and 8th of a month.

Also I think clicking on the placeholder when the popover is already open should close the popover.

On the devdocs page the placeholder dates are shown in US format. Will that change according to the user's locale? (assuming we have that information, and can derive a date format from it).

This may be totally out of scope, but I'm thinking about how someone could quickly enter slightly fuzzy dates. That is, 'anytime in 2018'. Without adding a bunch of new UI maybe the input fields could smart-parse the dates. For example, you enter the year '2018' into the 'from' and it expands to the start of 2018. Enter '2018' in the 'to' box and it expands to the end of 2018.

getdave commented:
> This is looking great and feels polished already. It makes a lot of sense to pull this out into a proper component.

Thank you! Pleased you think it is worthwhile 🎉 

> One slight thing I noticed is that the placeholder (the button you initially click to open the popover) changes width slightly as you change dates, making the popover move about. It's only a few pixels but you see it if you click between the 1st and 8th of a month.

Great spot! Will fix.

> ...clicking on the placeholder when the popover is already open should close the popover.

I just noticed this. Great spot. Will fix.

> On the devdocs page the placeholder dates are shown in US format. Will that change according to the user's locale? (assuming we have that information, and can derive a date format from it).

It [uses the `localize` HOC](https://github.com/Automattic/wp-calypso/blob/09a373fb0ddddd2ad75e4a29de28d87e04e35dc8/client/components/date-range/index.js#L360) from `i18n` to determine the date format as per your Calypso locale setting.

> This may be totally out of scope, but I'm thinking about how someone could quickly enter slightly fuzzy dates. That is, 'anytime in 2018'. Without adding a bunch of new UI maybe the input fields could smart-parse the dates. For example, you enter the year '2018' into the 'from' and it expands to the start of 2018. Enter '2018' in the 'to' box and it expands to the end of 2018.

Given that it's Friday...phase 2.0? :)

mattsherman commented:
Some critical styling issues in Firefox...

<img width="738" alt="screencapture at fri jan 11 08 32 53 est 2019" src="https://user-images.githubusercontent.com/2098816/51036725-fd84a480-157b-11e9-8fa3-c572ec684b40.png">

and Safari...

<img width="844" alt="screencapture at fri jan 11 08 34 17 est 2019" src="https://user-images.githubusercontent.com/2098816/51036729-02e1ef00-157c-11e9-977d-3cb505b02cb9.png">

getdave commented:
> Some critical styling issues in Firefox...
> and Safari...

⚒ Fixed and fixed.

![screen shot 2019-01-11 at 13 46 26](https://user-images.githubusercontent.com/444434/51037148-56b5fd80-15a7-11e9-8bce-0b77419424f6.png)

mattsherman commented:
@getdave I think you forgot to push your Firefox/Safari fixes. :-)

johngodley commented:
Mobile behaves a bit oddly too. On an iPad sized device you see the full popup, but clicking on a date also seems to trigger the next month

![ipad screen](https://user-images.githubusercontent.com/1277682/51038579-823ae700-15ab-11e9-96a5-2a691b0ace84.gif)

On a smaller device (iPhone) you only see one calendar and clicking changes both dates:

<img width="399" alt="ui_components_ _wordpress_com" src="https://user-images.githubusercontent.com/1277682/51038650-b31b1c00-15ab-11e9-9c3d-6d2bbbfb7214.png">

If you resize the display then it seems to behave in a responsive manner and puts the calendar underneath:

<img width="373" alt="ui_components_ _wordpress_com" src="https://user-images.githubusercontent.com/1277682/51038723-e9f13200-15ab-11e9-80f2-12236fa13c08.png">

Not sure why it doesn't do that originally with the small device.

getdave commented:
> Mobile behaves a bit oddly too. On an iPad sized device you see the full popup, but clicking on a date also seems to trigger the next month

Turns out this is to do with event handlers being fired for both `click` and `touch` events on touch-enabled devices.

See I’d like to discuss this with you
https://github.com/Automattic/wp-calypso/pull/27287/commits/406a5f6226761b5ef3ff17f35f5d42956880ea6e

I've contacted @enejb about it to see if he can explain _why_ he added touch support (the commit doesn't say) so that I can determine whether it can be removed.

mattsherman commented:
Tests fail for me (possible I have something configured incorrectly? New machine, but Calypso builds and runs without issue)...

```
 FAIL  client/components/date-range/test/index.js
  ● DateRange › should render

    expect(value).toMatchSnapshot()

    Received value does not match stored snapshot "DateRange should render 1".

    - Snapshot
    + Received

    @@ -5,13 +5,13 @@
          buttonRef={
            Object {
              "current": null,
            }
          }
    -     endDateText="01/06/2018"
    +     endDateText="31/05/2018"
          onTriggerClick={[Function]}
    -     startDateText="01/05/2018"
    +     startDateText="30/04/2018"
        />
        <Connect(Popover)
          className="date-range__popover is-dialog-visible"
          context={null}
          isVisible={false}
    @@ -26,15 +26,15 @@
              cancelButtonText="Cancel"
              onApplyClick={[Function]}
              onCancelClick={[Function]}
            />
            <DateRangeInputs
    -         endDateValue="01/06/2018"
    +         endDateValue="31/05/2018"
              endLabel="To"
              onInputBlur={[Function]}
              onInputChange={[Function]}
    -         startDateValue="01/05/2018"
    +         startDateValue="30/04/2018"
              startLabel="From"
            />
            <Localized(DatePicker)
              calendarViewDate={2018-05-01T00:00:00.000Z}
              className="date-range__popover-date-picker"

      56 | 	test( 'should render', () => {
      57 | 		const wrapper = shallow( <DateRange moment={ moment } /> );
    > 58 | 		expect( wrapper ).toMatchSnapshot();
         | 		                  ^
      59 | 	} );
      60 | 
      61 | 	describe( 'Date range clamping', () => {

      at Object.toMatchSnapshot (client/components/date-range/test/index.js:58:21)

  ● DateRange › Trigger element › should render trigger with appropriate default date range of minus one month from today's date

    expect(received).toEqual(expected)

    Expected value to equal:
      ObjectContaining {"endDateText": "01/06/2018", "startDateText": "01/05/2018"}
    Received:
      {"buttonRef": {"current": null}, "endDateText": "31/05/2018", "onTriggerClick": [Function wrapper], "startDateText": "30/04/2018"}

    Difference:

    - Expected
    + Received

    - ObjectContaining {
    -   "endDateText": "01/06/2018",
    -   "startDateText": "01/05/2018",
    + Object {
    +   "buttonRef": Object {
    +     "current": null,
    +   },
    +   "endDateText": "31/05/2018",
    +   "onTriggerClick": [Function wrapper],
    +   "startDateText": "30/04/2018",
      }

      140 | 			};
      141 | 
    > 142 | 			expect( dateRangeTrigger.props() ).toEqual( expect.objectContaining( expected ) );
          | 			                                   ^
      143 | 		} );
      144 | 
      145 | 		test( 'should update trigger props to match currently selected dates', () => {

      at Object.toEqual (client/components/date-range/test/index.js:142:39)

  ● DateRange › Input elements › should not update date selection on input change event

    expect(received).toEqual(expected)

    Expected value to equal:
      "01/06/2018"
    Received:
      "31/05/2018"

      389 | 			wrapper.instance().handleInputChange( endDate, 'End' );
      390 | 
    > 391 | 			expect( wrapper.state().endDate.format( 'DD/MM/YYYY' ) ).toEqual(
          | 			                                                         ^
      392 | 				fixedEndDate.format( 'DD/MM/YYYY' )
      393 | 			);
      394 | 		} );

      at Object.toEqual (client/components/date-range/test/index.js:391:61)

 › 1 snapshot failed.
Snapshot Summary
 › 1 snapshot failed from 1 test suite. Inspect your code changes or press `u` to update them.

Test Suites: 1 failed, 1 total
Tests:       3 failed, 28 passed, 31 total
Snapshots:   1 failed, 1 total
Time:        1.522s
Ran all test suites matching /client\/components\/date-range\/test/i.

Active Filters: filename /client/components/date-range/test/
```

mattsherman commented:
> > Mobile behaves a bit oddly too. On an iPad sized device you see the full popup, but clicking on a date also seems to trigger the next month
> 
> Turns out this is to do with event handlers being fired for both `click` and `touch` events on touch-enabled devices.
> 
> See I’d like to discuss this with you
> [406a5f6](https://github.com/Automattic/wp-calypso/commit/406a5f6226761b5ef3ff17f35f5d42956880ea6e)
> 
> I've contacted @enejb about it to see if he can explain _why_ he added touch support (the commit doesn't say) so that I can determine whether it can be removed.
> 
> > On a smaller device (iPhone) you only see one calendar...
> 
> This is intentional. By default screens smaller than 480px get this experience because 9 times out of 10 there isn't screen real-estate on a real device to render either:
> 
>     1. 2-horiztonal
> 
>     2. 2-stacked
> 
> 
> You can customise what determines the 1-up vs 2-up layout via the `numMonths` prop so if you need to force 2-up or 1-up or just alter the breakpoint you can easily do that.

Forcing 2-up seems to just be asking for problems -- what are the scenarios where a dev would need to override the breakpoint handling of the component to force this? Or, force a 1-up when the viewport size will support 2-up?

mattsherman commented on client/components/date-range/trigger.js:
```diff
@@ -0,0 +1,53 @@
+				<span className="date-range__trigger-btn-text">
+					{ props.startDateText }
+					&nbsp; - &nbsp;
+					{ props.endDateText }
```
Is this formatting (start date, dash, then end date) locale agnostic? I'm not sure if some locales would use a different separator or even put them in different order?

If using dash, should use proper `&ndash;`.

getdave replied:
Hmmm that _is_ a good point. Options

1. Give consumer a way to format the entire string `startDate + endDate`
2. Give consumer a way to choose separator
3. Find a way to make the separator locale aware (if that is a thing)

getdave replied:
Found this which also suggests making it open to translators

https://stackoverflow.com/questions/10449487/whats-a-i18n-friendly-way-to-show-a-date-range

getdave replied:
Resolve based on comments here `p1547464508137900-slack-calypso` and existing code [here](https://github.com/Automattic/wp-calypso/blame/master/client/my-sites/stats/stats-date-picker/index.jsx#L81)

getdave commented on client/components/date-range/header.js:
```diff
@@ -0,0 +1,51 @@
+					primary
+					compact
+				>
+					<Gridicon icon="checkmark" />
```
Have requested @drw158 who reviewed https://github.com/Automattic/wp-calypso/pull/25762.

getdave replied:
Whilst we're waiting on design I've been looking at the existing Components and Blocks and have found these component which use icons

* [CalendarButton](http://calypso.localhost:3000/devdocs/blocks/calendar-button)
* [LikeButton](https://github.com/Automattic/wp-calypso/pull/29938#pullrequestreview-191737466)
* [CommentButton](http://calypso.localhost:3000/devdocs/blocks/comment-button)
* [FollowButton](http://calypso.localhost:3000/devdocs/blocks/follow-button)
* [ButtonGroup](http://calypso.localhost:3000/devdocs/design/button-group) demos
* [Button](http://calypso.localhost:3000/devdocs/design/button) demos
* [SplitButton](http://calypso.localhost:3000/devdocs/design/split-button) demos

drw158 replied:
Using icons in buttons is fine. In this case, I'd removed the check icon just to be consistent with the cancel button.

getdave replied:
Have removed ✔️ icon on `Apply` button as requested.

getdave commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,526 @@
+	 * @param  {Date|MomentJSDate} date the date to be converted
+	 * @return {String}      the date as a formatted locale string
+	 */
+	dateToHumanReadable( date ) {
```
Agreed 👍

getdave replied:
Fixed. Now `formatDateToLocale`

getdave commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,526 @@
+	 * @param  {MomentJSDate} momentDate a momentjs date object to convert
+	 * @return {DATE}            the converted JS Date object
+	 */
+	momentDateToNative( momentDate ) {
```
Agreed 👍

getdave replied:
Fixed

getdave commented on client/components/date-range/inputs.js:
```diff
@@ -0,0 +1,81 @@
+					<FormLabel htmlFor="startDate">{ this.props.startLabel }</FormLabel>
+					<FormTextInput
+						id="startDate"
+						name="startDate"
```
Great catch. It could cause a problem. For some reason I thought react would prefix it in virtual dom but clearly I was way off!

Thinking of using https://lodash.com/docs/4.17.11#uniqueId to generate per instance Ids.

getdave replied:
I've utilised `uuid` to generate a per instance ID which I use to generate unique `input` IDs. 

![screen shot 2019-01-14 at 09 54 55](https://user-images.githubusercontent.com/444434/51106093-b8ae7700-17e2-11e9-8983-fdb65ebd516a.png)

getdave commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,526 @@
+	renderPopover() {
+		const popoverClassNames = classNames( {
+			'date-range__popover': true,
+			'is-dialog-visible': true, // forces to render when inside a modal. Is there a way to detect this from global state?
```
Yep. Same here. The comment was from the original PR where I obviously had also felt uncomfortable.

Is there a guru on such things I could ask?

If not I will have to do some more digging.

mattsherman replied:
Hmm... not sure who would be most familiar with dialog and popover currently. I actually wrote the first versions of both, but believe they have been modified a good deal since I last touched them. My implementation was before we had global state, etc.

getdave commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,526 @@
+	 * Closes the popover
+	 * Note this does not commit the current date state
+	 */
+	closePopover() {
```
We are looking at: `open/close` vs `open/cancel`. 

Worth noting that `closePopover` is also called when you click "Apply" and so it isn't tied to either applying or canceling per se. As a result I'd still say that `close` is a nice agnostic name which marries well with its antonym `open`.

mattsherman replied:
Ah, in that case, I agree!

getdave commented on client/components/date-picker/style.scss:
```diff
@@ -277,3 +277,25 @@ $date-picker_nav_button_size: 20px;
+
+.date-picker.date-picker--has-range .DayPicker-Day.DayPicker-Day--range-start {
+	.date-picker__day {
+		background: $blue-medium;
```
Thanks for the great resource. I'll be sure to convert those to CSS vars

getdave replied:
Converted!

getdave commented on client/components/date-picker/index.jsx:
```diff
@@ -7,8 +7,9 @@
 import PropTypes from 'prop-types';
 import DayPicker from 'react-day-picker';
-import { noop, merge, map, filter, get } from 'lodash';
+import { noop, map, filter, get, merge } from 'lodash';
```
Will fix that. It's a long story...

getdave replied:
Fixed

getdave commented:
> Forcing 2-up seems to just be asking for problems -- what are the scenarios where a dev would need to override the breakpoint handling of the component to force this? Or, force a 1-up when the viewport size will support 2-up?

@mattsherman As discussed on Slack, we're going to drop the `numMonths` overide prop and force people to stick to our choice of when UI switches from 1up to 2up. This doesn't mean in the future that people won't find a need to enable this to be amended, but at the moment we have no clear need for it and so we should keep the API as simple as possible.

mattsherman commented:
The component "twitches" slightly when selecting dates with a slightly different text format width. This is distracting and feel unpolished!

![date range twitch](https://user-images.githubusercontent.com/2098816/51054690-9b439800-15ab-11e9-9756-6f20fd937025.gif)

getdave commented on client/components/date-range/style.scss:
```diff
@@ -0,0 +1,126 @@
+.date-range__span {
+	margin: 0 8px 0 0;
+	font-size: 0.9em;
+	color: $gray;
```
Done

getdave commented:
> The component "twitches" slightly when selecting dates with a slightly different text format width. This is distracting and feel unpolished!

This has been raised before by @johngodley. As a result, I decided to look into it again as it's clearly an issue that people notice.

I've decided to resolve with a fixed width on the button which accommodates the text length changing.

spen commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,526 @@
+		let startDate;
+		let endDate;
+
+		endDate = ! isNil( this.props.selectedEndDate )
```
We can invert the ternaries here to avoid having to negate the checks:

```js
endDate = isNil( this.props.selectedEndDate )
  ? this.props.moment()
  : this.props.moment( this.props.selectedEndDate );
```

getdave replied:
Yeh that is clearer - thanks.

getdave replied:
Done

getdave commented:
> Tests fail for me (possible I have something configured incorrectly? New machine, but Calypso builds and runs without issue)...

@mattsherman I've [updated the tests](https://github.com/Automattic/wp-calypso/pull/29938/commits/6d601829b6ce8fc00d43289c7606db30a08691ef) to use ISO and US date formattingby default. I am 100% ✅. Could you possibly confirm whether or not this has also fixed for you?

spen commented on client/components/date-range/style.scss:
```diff
@@ -0,0 +1,129 @@
+
+
+.date-range__trigger-btn-text {
+	width: 12.5em; // required to avoid content "jumping" as date text changes
```
This change tests well and I think I do prefer this over using a monospace font (we shouldn't be forced to select a font based on this problem).
What do you think about changing this to `min-width` though? I noticed that changing the characters to be wider manually resulted in a line break. Of course these actual values couldn't exist, but there may be some valid dates that are equally as wide:

<img width="222" alt="screenshot 2019-01-14 at 15 14 03" src="https://user-images.githubusercontent.com/4335450/51121480-794a4f80-180f-11e9-9a21-eef463611836.png">

getdave replied:
> This change tests well and I think I do prefer this over using a monospace font (we shouldn't be forced to select a font based on this problem).

👍 

> What do you think about changing this to min-width though? 

Yes `min-width` would be safer just in case some random value we haven't tested causes it to expand.

>  I noticed that changing the characters to be wider manually resulted in a line break

I was trying to find as many horrible combinations as possible - I knew I'd miss one! Thanks for being so thorough. I'll tweak again slightly wider and then move to `min-width`

getdave replied:
✅ Done

spen commented on client/components/date-range/style.scss:
```diff
@@ -0,0 +1,129 @@
+		display: flex;
+	}
+
+	.DayPicker-Month {
```
Just pointing this out so that we either make use of it or 🔪it out before merge :)

getdave replied:
Note to self: this refers to need to remove unused selector

getdave replied:
Fixed

spen commented on client/components/date-range/style.scss:
```diff
@@ -0,0 +1,129 @@
+	display: flex;
+	align-items: baseline;
+}
+
```
Nit: extra new-line here

getdave replied:
Fixed all spacing to 1 line

getdave commented on client/components/date-range/trigger.js:
```diff
@@ -0,0 +1,67 @@
+			return this.props.triggerText( startDateText, endDateText );
+		}
+
+		return this.props.translate( '%(startDateText)s - %(endDateText)s', {
```
Yes, there is potential for this to be the case (apparently).

getdave replied:
Yes, apparently they do. Two people have said this now.

spen replied:
Interesting!

getdave commented on client/components/date-range/trigger.js:
```diff
@@ -0,0 +1,67 @@
+	dateRangeText() {
+		const { startDateText, endDateText } = this.props;
+
+		if ( this.props.triggerText ) {
```
> No need to do it here as there's no real benefit

In this context what is "this"?

The aim here is to only invoke the `triggerText ` function if it has been passed in as a prop, _otherwise_ fallback to a default. 

What am I missing?

> Lodash has the invoke function

Thanks for this. I always missed that from PrototypeJS back in the day.

getdave replied:
👍

getdave commented on client/components/date-range/inputs.js:
```diff
@@ -0,0 +1,92 @@
+	 * @param  {string} startOrEnd one of "Start" or "End"
+	 * @return {function}            the partially applied function ready to recevie event data
+	 */
+	handleInputEvent = ( eventType, startOrEnd ) => e => {
```
No reason other than remaining as `DRY` as possible. That said, it's probably clearer to have two distinct methods even if they share the same logic. I'll look at a quick refactor here.

getdave replied:
Fixed

getdave commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,526 @@
+		};
+
+		// Bind event handlers
+		bindAll( this, [
```
Ah so this is formal now. Great. Arrow functions all the way then!

spen replied:
🏹💚

getdave replied:
Done

spen commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,526 @@
+		const firstSelectableDate = ! isNil( this.props.firstSelectableDate )
+			? this.props.moment( this.props.firstSelectableDate )
+			: undefined;
+		const lastSelectableDate = ! isNil( this.props.lastSelectableDate )
```
I might be misinterpreting the usage of `isNil` here, but would the usage of `has` work for us here?: 
```js
const lastSelectableDate = has( this.props, 'lastSelectableDate' ) && this.props.moment( this.props.firstSelectableDate );
```
If so we can avoid a double negative (`! isNil` is equal to `is not nil`) and better show intent.
Just a thought though, I could be missing something there.

getdave replied:
Seems to work. I used the original because I thought we'd need `undefined` explicitly to "turn off" the feature in the `DatePicker` rather than `false` which `has()` returns. But your solution works just fine and is much clearer as to the intent. I was never comfortable with a ternary condition resulting in `undefined`. It just looked wrong.

Fixed.

spen commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,526 @@
+	 * @param  {Moment|Date} options.dateTo   the end of the date range
+	 * @return {Moment}                  the date clamped to be within the range
+	 */
+	clampDateToRange( date, { dateFrom, dateTo } ) {
```
`clampDateToRange` looks like it could live as a standalone util function as it doesn't seem to benefit from being a member of the component.
That said, this definitely isn't a blocker as there's also no harm in it being a member method.

getdave replied:
Will have a 🤔

getdave commented on client/components/date-range/trigger.js:
```diff
@@ -0,0 +1,67 @@
+ * Internal dependencies
+ */
+import Button from 'components/button';
+import Gridicon from 'gridicons';
```
Done

spen replied:
Thanks 🙇

spen commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,526 @@
+		let config = {};
+
+		if ( firstSelectableDate ) {
+			config = {
```
Rather than reassigning the value of `config` we could use regular property assignment here:

```js
const config = {};

if ( firstSelectableDate ) {
  config.before = ...
}
if ( lastSelectableDate ) {
  config.after = ...
}
```

We could even go sort of [point-free](https://wiki.haskell.org/Pointfree) with:

```js
return [ {
  ...firstSelectableDate && { before: .... }, 
  ...lastSelectableDate && { after: .... },
} ]
```

But I still think the 2 `if`s read better

getdave replied:
I must have been tired when writing this code! I'll go with the simple `if/else` as it's much clearer.

spen replied:
I like that you were thinking of immutability though ;)

drw158 commented:
I don't think using a monospace font is a great idea. The reasoning behind using the monospace font is a technical problem, not really a visual one. I'd say don't worry about, or find a way to make the pop up remain still.

getdave commented on client/components/date-range/inputs.js:
```diff
@@ -0,0 +1,92 @@
+	 * @return {function}            the partially applied function ready to recevie event data
+	 */
+	handleInputEvent = ( eventType, startOrEnd ) => e => {
+		const val = e.target.value;
```
Fixed and noted.

getdave commented:
# Update

The component has been updated  with:

✅ all CR comments addressed
✅ standard size by default, "compact" size via `prop` option `isCompact`
✅ tests refactored (waiting on @mattsherman to confirm)
✅ "jumping on date select" fixed
✅  a11y improvements addressed (where feasible)

![screen capture on 2019-01-15 at 10-44-33](https://user-images.githubusercontent.com/444434/51175642-c92e2280-18b2-11e9-9164-674053c9cf15.gif)

spen commented on client/components/date-range/test/index.js:
```diff
@@ -0,0 +1,632 @@
+
+			wrapper.instance().handleInputBlur( '05/28/2018', 'End' );
+
+			expect( dateToLocalString( wrapper.state().endDate ) ).toEqual(
```
What's the reason for the `endDate` test to be different to the `startDate` test in terms of comparing the formatted values?

getdave replied:
Human error. Fixed.

spen commented on client/components/date-range/test/index.js:
```diff
@@ -0,0 +1,632 @@
+				/>
+			);
+
+			wrapper.instance().handleInputBlur( '04/20/2018', 'Start' );
```
Just a thought here - no need to update.
When reading through tests in a large test file I like to see inputs defined locally to the tests where possible or for constants to be used so that I don't need to scroll up and down the file for context.
Here for example, I'd have to remind myself of the significance of `'04/20/2018'` in relation to `momentStartDate` by scrolling up (which interrupts my reading of the test cases).

We can be more redundant with tests as their performance isn't as important as it is in a product so redefining moments and such is fine if it helps make the intent of the test case more clear:

```js
const startDate = '04/20/2018';
const startDateMoment = moment( startDate );
const wrapper = shallow(
  <DateRange
    selectedStartDate={ momentStartDate }
    selectedEndDate={ momentEndDate }
    moment={ moment }
  />
);

wrapper.instance().handleInputBlur( startDate, 'Start' );
```

Just wanna hammer home that this isn't a change request as whats here is great, just a personal thought/opinion :)

getdave replied:
I'm glad you raised this. I've been thinking about this and having done a lot of reading about testing best practices recently I'm convinced that we should avoid abstracting tests as much as possible in favour of clarity and simplicity.

I'll look into a refactor of this - shouldn't take long.

getdave replied:
I totally agree with this. Having done a lot more reading about testing best practices, I'm also convinced that it's better to have redundancy in rests and be clearer.

> no need to update.

Agree. I don't think I will refactor these now but I will keep in mind for next time.

spen commented on client/components/date-range/test/index.js:
```diff
@@ -0,0 +1,632 @@
+					'buttonRef',
+					'onTriggerClick',
+					'triggerText',
+				].sort()
```
Could we not just define these keys ready-sorted?

spen replied:
Actually, I see the value of just allowing `[].sort()` to handle the sorting as it's the same `sort`ing as in done in `Object.keys( props ).sort();` above 👍

getdave replied:
Was trying to avoid a situation whereby order the order of props in the component would cause tests to fail.

spen commented on client/components/date-range/inputs.js:
```diff
@@ -0,0 +1,97 @@
+						/>
+					</div>
+					<div className="date-range__date-input date-range__date-input--to">
+						<FormLabel htmlFor={ this.endDateID }>{ this.props.endLabel }</FormLabel>
```
One thing that could be an issue here is that the labels aren't translated in the default props so the responsibility to do so does fall on the implementation - that's fine as long as we're aware of it 👍

getdave replied:
Good spot. I've wrapped default props in `translate` with context. If someone is overriding then they will need to translate the input yes.

spen commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,511 @@
+				// For first date selection only: "cache" previous dates
+				// just in case user doesn't "Apply" and we need to revert
+				// to the original dates
+				if ( ! this.state.staleDatesSaved ) {
```
Slightly nitty question here (not a blocker): Should we use `previousState. staleDatesSaved` here for the sake of consistency? Or is there purpose in using `this.state` here?

getdave replied:
Fixed.

spen commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,511 @@
+		// By default check
+		// 1. Looks like a valid date
+		// 2. after 01/01/1970 (avoids bugs when really stale dates are treated as valid)
+		let validations = date.isValid() && date.isSameOrAfter( epoch );
```
`isValid` would better represent a `boolean` value, `validations` would suggest to me either a collection of validation objects (to be used in showing a validation message, for example).

It looks like we could be making use of early returns too to avoid the need for `validations` completely , something like so:
```js
if ( ! date.isValid() || ! date.isSameOrAfter( epoch ) ) {
  return false;
}

if ( firstSelectableDate &&  date.isBefore( firstSelectableDate ) {
  return false;
}

if ( lastSelectableDate &&  date.isAfter( lastSelectableDate ) {
  return false;
}

return true;
```

That allows us to skip some extra checking and reassigning by eagerly rejecting the validation.
As a bonus It also sets us up nicely for the possibility of adding specific validation errors, if the need arises:

```js
if ( ! date.isValid() || ! date.isSameOrAfter( epoch ) ) {
  return new Error( 'some message relating to the above condition' );
}

if ( firstSelectableDate &&  date.isBefore( firstSelectableDate ) {
  return new Error( 'some message relating to the above condition' );
}

if ( lastSelectableDate &&  date.isAfter( lastSelectableDate ) {
  return new Error( 'some message relating to the above condition' );
}

// returning false or just nothing at all would suggest no validation issues here
return false;
```

spen replied:
I do think it's easier as a reviewer to spot stuff like this than it is while writing the code itself. It's the same reason I often find myself looking back at my own code with confusion 😄

spen commented on client/components/date-range/header.js:
```diff
@@ -0,0 +1,49 @@
+	static defaultProps = {
+		onApplyClick: noop,
+		onCancelClick: noop,
+		applyButtonText: 'Apply',
```
There seems to be an issue with translations here and it's rather easy to miss:

`DateRangeHeader` accepts it's props through `DateRange`, so in order for `applyButtonText` and `cancelButtonText` to be translated, the dev implementing `DateRange` would have to be sure to do something like so:

```jsx
<DateRange
  applyButtonText={ translate( 'Apply' ) }
  closeButtonText={ translate( 'Close' ) }
  ...
/>
```

Failing to do so would mean falling back to these untranslated defaults. The trouble is here that we're quite quiet about this requirement in that A) these props are not marked as required in this component & B) They're not advertised at all in `DateRange`, which is their real entry point.

I see a couple of options here...
We could keep these props as optional, but we take care of translating the default values, like so:

```jsx
class DateRangeHeader extends Component {
  ...

  static defaultProps = {
    onApplyClick: noop,
    onCancelClick: noop,
  };

  ...

  render() {
    const { translate } = this.props;
    return (
      <div className="date-range__popover-header">
        <Button>
          { this.props.cancelButtonText || translate( 'Cancel' ) }
        </Button>
        ...
      </div>
    );
  }
}

export default localize( DateRangeHeader );
```

Or we could remove the defaults and mark these props as required in both the `DateRangeHeader` and `DateRange` components.
I'd personally go for the first pattern as it alleviates the extra need to define these labels, yet still handles translations.

Which ever we go for, lets apply the same treatment to the other subcomponents.

BTW, I'm happy to scratch this up as work for another PR so as not to block this current work.

getdave replied:
Yep I'm going to bank that pattern. It's the right approach. Can't rely too much on `defaultProps`.

Fixed

spen commented on client/components/date-range/docs/example.jsx:
```diff
@@ -0,0 +1,84 @@
+					<DateRange lastSelectableDate={ now } />
+				</Card>
+
+				<h3>Select only future dates</h3>
```
Liking the coverage of the variations 👌

spen commented on client/components/date-range/README.md:
```diff
@@ -0,0 +1,50 @@
+import DateRange from 'components/date-range';
+
+export default class extends React.Component {
+	<DateRange />
```
This example is quite bare and doesn't tell us much, it'd be cool to see some elaboration (perhaps as follow up work) to show some of the more custom usages in the future 👌

spen replied:
Also, I've just noticed that this usage isn't valid as `<DateRange />` isn't part of the render method.
Let's either move it into a render method or (preferred) just make this in to a functional component.

getdave replied:
> This example is quite bare and doesn't tell us much, it'd be cool to see some elaboration (perhaps as follow up work) to show some of the more custom usages in the future

The particular "default" example or the other ones? Was aiming to keep it reasonably simple and encourage others to document use cases as they found them, but happy to update in future as I find them!

> Let's either move it into a render method or (preferred) just make this in to a functional component.

Fixed

spen replied:
> Was aiming to keep it reasonably simple and encourage others to document use cases as they found them

That sounds good and fair  👍

spen replied:
> The particular "default" example or the other ones?

I just meant the one in the readme, the actual DevDocs examples are great :)

spen commented on client/components/date-range/README.md:
```diff
@@ -0,0 +1,50 @@
+DateRange
+==========
+
+React component used to display a Date Range.
```
Nit - lets add `display and select` here as selecting the range is what really makes this component :)

getdave replied:
Fair. Had already updated but have re-updated again to reinstate the `display` bit as well.

spen commented on client/components/date-picker/index.jsx:
```diff
@@ -203,24 +215,39 @@ class DatePicker extends PureComponent {
+				fromMonth={ this.props.fromMonth }
+				toMonth={ this.props.toMonth }
 				onDayClick={ this.setCalendarDay }
-				onDayTouchStart={ this.setCalendarDay }
```
Are we losing anything here by removing these props?

getdave replied:
This is one for @enejb to help with.

Please see https://github.com/Automattic/wp-calypso/pull/27287/commits/406a5f6226761b5ef3ff17f35f5d42956880ea6e

Essentially if we leave this in, each time you "touch" a day it triggers the `setCalendarDay` multiple times because the event fires for both touch and click. 

The commit isn't 100% about why this was added apart from "for a11y reasons". However we know click event are fired on touch devices so without a clear reasoning I'm inclined to say it needs to be removed otherwise the range selection logic doesn't work.

spen replied:
Got ya, thanks for the explanation :)

spen commented on client/components/date-picker/index.jsx:
```diff
@@ -203,24 +215,39 @@ class DatePicker extends PureComponent {
+		}
+
+		const rootClassNames = classNames( {
+			'date-picker': true,
```
A tip for classNames that are always true, we can pass a regular strings as arguments to `classNames` like so:

```js
const rootClassNames = classNames(
  'date-picker',
  `date-picker--${ this.props.numberOfMonths }up`,
  {
    somethingConditional: someCondition,
  }
);

(Not a blocker)

getdave commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,511 @@
+			startDate: startDate,
+			endDate: endDate,
+			staleDatesSaved: false,
+			// this need to be independent from startDate because we must independently validate them
```
Fixed

getdave commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,511 @@
+			textInputEndDate: this.formatDateToLocale( endDate ),
+		};
+
+		// Ref to the Trigger <button> used to position the Popover	component
```
Not following this sorry 😕

spen replied:
Sorry, my bad I meant the tab character in the comment here:

<img width="618" alt="screenshot 2019-01-15 at 15 43 59" src="https://user-images.githubusercontent.com/4335450/51191485-6487bd80-18dc-11e9-8e51-187bf0f255e5.png">

getdave replied:
No probs. Fixed.

getdave commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,511 @@
+	 * @return {ReactComponent} the Popover component
+	 */
+	renderPopover() {
+		const popoverClassNames = classNames( {
```
Fixed

beaulebens commented:
How does this differ/do we need a new date range picker, when we already have one as used in the Activity Log?

<img width="372" alt="screen shot 2019-01-15 at 5 17 18 pm" src="https://user-images.githubusercontent.com/108942/51218413-89366200-18e9-11e9-9568-f713a21e97d6.png">

beaulebens commented:
I'll cc in @enejb and I think @keoshi who worked on the design of the Activity Log one for reasoning/background on decisions made there. I don't personally have a preference which direction we go, other than not having 2 date range pickers :)

getdave commented:
@keoshi Thanks so much for taking the time to review this. Responses below

> TO date not correct:

Fixed.

> Position of the Cancel and Apply buttons:

Move to top and unified with the info bar. Stacks to 1 column on smaller screens.

> Date range styles:

I've merged in the styles from the Activity log and made some consistency improvements. Should look the same now.

> :hover styles

I've added both `:hover` and `:focus` styles to each day as a very subtle `neutral` shade. Hopefully this works?

> Nudging to the left on selection:

This issue has been a source of problems all the way through. Thanks for tip on relative position. I've bumped up the `min-width` as a quick fix which stops the immediate issue. Longer term a better solutions needs to be found.

> View one/two months at a time:

I think you are right. @mattsherman wasn't keen on us having a prop for this unless we had a good reason in order to keep the API simple. However, now you've mentioned it _and_ we have a clear use-case in the Activity log I think there's an argument to reinstate. But not now given the MVP and timelines.

keoshi commented:
@getdave Thank you so much for all the work and effort put into this. The new range selection styles look stellar in my opinion!!

getdave commented:
@enejb With a bit of work @spen and I have resolved the issue with the Activity Log DatePicker date selection. See below on iOS simulator which shows both have been fixed

### Activity Log 

![screen capture on 2019-01-21 at 11-56-19](https://user-images.githubusercontent.com/444434/51473988-c8473600-1d75-11e9-8abc-b7f6808bd5c7.gif)

### New Component (this PR)

![screen capture on 2019-01-21 at 11-55-31](https://user-images.githubusercontent.com/444434/51473991-ca10f980-1d75-11e9-9808-03f27ab08a65.gif)

spen commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,665 @@
+			} );
+		}
+
+		startDate = isNil( this.props.selectedStartDate )
```
This seems to be more complicated than it need be, with some indirection that might be avoidable:

For starters, here are some things to consider:
```js
// isNil === null or undefined.
// isNull === null.
// NO_DATE_SELECTED_VALUE = null;
```
so 
```js
endDate = isNil( this.props.selectedEndDate )
	? NO_DATE_SELECTED_VALUE
	: this.props.moment( this.props.selectedEndDate );

// essentially does this:
if ( selectedEndDate === null ) {
  endDate = null;
} else if ( selectedEndDate === undefined ) {
  endDate = null;
} else {
  endDate = this.props.moment( this.props.selectedEndDate )
}
```
We then check for `! isNull( endDate )` but we already know that it's either `null` or a `moment` because we coerced any `undefined`s to `NO_DATE_SELECTED_VALUE`/`null`

Because of that, I think we could merge the two conditionals in to:

```js
const endDate = isNil( selectedEndDate )
	? NO_DATE_SELECTED_VALUE
	: this.clampDateToRange(
		this.props.moment( selectedEndDate ), 
		{
			dateFrom: firstSelectableDate,
			dateTo: lastSelectableDate,
		}
	);
```

spen replied:
With the above we could make a couple of extra tweaks that would allow us to avoid using `let` for `endDate` and `startDate`, though I'm not suggesting we do (not worth the time).
We could theoretically do so by making and using a util like so 
```js
startDate: getEarliest( startDate, endDate ),
endDate: getLatest( startDate, endDate ),
initialStartDate: getEarliest( startDate, endDate ),
initialEndDate: getLatest( startDate, endDate ),
```

getdave replied:
Thanks. I never liked this but was focusing attention elsewhere.

Have simplified to a degree with the MVP and time restrictions in mind.

spen replied:
Awesome, thanks @getdave !

spen commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,665 @@
+import DateRangeTrigger from './trigger';
+import Button from 'components/button';
+import Gridicon from 'gridicons';
+const NO_DATE_SELECTED_VALUE = null;
```
A nit here, lets bump this down for a little added clarity:
```suggestion

/**
 * Module variables
 */
const NO_DATE_SELECTED_VALUE = null;
```

getdave replied:
Fixed

spen commented on client/components/date-range/index.js:
```diff
@@ -0,0 +1,665 @@
+/**
+ * Internal dependencies
+ */
+import { localize, moment } from 'i18n-calypso';
```
Nit: 

`import { localize, moment } from 'i18n-calypso';`
and
`import Gridicon from 'gridicons';`
should live under `External dependencies`

getdave replied:
You keep having to remind me of this! Sorry. 

Fixed.

spen replied:
NP! Easy to forget when there's more important stuff to fix :P

getdave commented:
Rebasing and force-push'ing now. Do not touch!

getdave commented:
🚢'd

Thanks to all who contributed their time and effort

- @spen 
- @enejb 
- @mattsherman 
- @keoshi 
- @drw158 
- @johngodley
